runtime/src/commonMain/kotlin/pbandk/json/JsonWireMarshaller.kt

```diff
@@ -0,0 +1,183 @@
+package pbandk.json
+
+import kotlinx.serialization.json.*
+import pbandk.*
+import pbandk.wkt.*
+import kotlin.Any
+import kotlin.reflect.KProperty1
+
+class JsonWireMarshaller(private val jsonConfig: JsonConfig) :
+    WireMarshaller {
+    private val json: Json =
+        Json(
+            JsonConfiguration.Stable.copy(
+                prettyPrint = true
+            )
+        )
+    private val jsonValueMarshaller = JsonValueMarshaller(jsonConfig)
+    private var currentMessage: JsonElement? = null
+
+    fun toJsonString(): String = currentMessage?.let { json.stringify(JsonElementSerializer, it) }.orEmpty()
+
+    fun toJsonElement() = currentMessage ?: error("Must call writeMessage() before toJsonElement()")
+
+    override fun <T : Message> writeMessage(message: T) {
+        check(currentMessage == null) { "JsonWireMarshaller can't be reused with multiple messages" }
+        currentMessage = when (message) {
+            // Wrapper types use the same JSON representation as the wrapped value
+            // https://developers.google.com/protocol-buffers/docs/proto3#json
+            is DoubleValue -> writeWrapperValue(message, message.value)
+            is FloatValue -> writeWrapperValue(message, message.value)
+            is Int64Value -> writeWrapperValue(message, message.value)
+            is UInt64Value -> writeWrapperValue(message, message.value)
+            is Int32Value -> writeWrapperValue(message, message.value)
+            is UInt32Value -> writeWrapperValue(message, message.value)
+            is BoolValue -> writeWrapperValue(message, message.value)
+            is StringValue -> writeWrapperValue(message, message.value)
+            is BytesValue -> writeWrapperValue(message, message.value)
+            // Other well-known types with special JSON encoding
+            is Timestamp -> jsonValueMarshaller.writeString(
+                Util.timestampToString(
+                    message
+                )
+            )
+            // All other message types
+            else -> writeMessageObject(message)
+        }
+    }
+
+    private fun writeWrapperValue(message: Message, value: Any) =
+        jsonValueMarshaller.writeValue(value, message.fieldDescriptors[0].type)
+
+    private fun <T : Message> writeMessageObject(message: T): JsonObject {
+        val jsonContent: MutableMap<String, JsonElement> = linkedMapOf()
+
+        for (fd in message.fieldDescriptors) {
+            @Suppress("UNCHECKED_CAST")
+            val value = (fd.value as KProperty1<T, *>).get(message)
+
+            if (value == null && fd.oneofMember) continue
+            if (!fd.oneofMember && !jsonConfig.outputDefaultValues && fd.type.isDefaultValue(value)) continue
```

Comment:
Does this mean we won't have empty strings in our response now? 🎉

Reply:
Yeah. By default (when `jsonConfig.outputDefaultValues == false`) empty strings, zero values, false values, and empty lists will be omitted from the response.

One idea for a future enhancement is to add a setting along the lines of `jsonConfig.outputDefaultValuesForFields = fieldList` to allow you to choose specific fields whose default values will be output. I think we had talked about something like this in the past.

Reply:
Yeah, we previously explicitly allowed `total_count` to return as `0`, which was useful.